docs/proposals/apiserver-ha.md

```diff
@@ -0,0 +1,152 @@
+| apiserver-1 |      | apiserver-2 |        | apiserver-3 |         
+|-------------|      |-------------|        |-------------|   
+      |                     |                     |
+      |                     |                     |
+|---------------------------------------------------------|             
+|  Nginx/HAProxy, or client randomly select an endpoint   |         
+|---------------------------------------------------------|
+       |                     |                     |
+       |                     |                     |
+ |------------|       |------------|        |-------------|              
+ |  scheduler |       | controller |        |    node     |      
+ |------------|       |------------|        |-------------|          
+```
+
+### Initional Implementation step
+
+#### 1. Introduce a knob to enable quorum read
+
+There are two cases where we may read old data if we have multi API server:
+
+1) When a etcd member is isolated from etcd cluster, we can still read from
+it(but all the write request sent to the isolated member will failed). If
+unfortunately nginx forward some read request to the isolated member,
+those read request will get old data.
+
+2) Consider the following case:
+
+A write request(e.g. ?key=hello&value=world) was forward to apiserver-1
+and then forward to etcd-1, etcd-1 will return 200 indicating the write
+request was succeeded if the data has been written to a majority of members.
+(for example, written to etcd-1 and etcd-2) Almost immediately, a read
+request(?key=hello) was forward to etcd-3, if API server doesn't enable quorum
+read, this read request will failed(key not found) since the data has not been
+synced to etcd-3.
+
+So, we must introduce a knob to enable quorum read if we have multi API server.
+Since quorum read will make read request less effective, user could disenable it
+if they just want a single API server.
+
+
+#### 2. Default kubernetes service/endpoints
+
+When API server start, it will create a kubernetes service, and will periodically
+reconcile it. Currently, API server assume there is only one endpoint(benckend)
+for the service, every apiserver will constantly overwrites the etcd key
+“/registry/services/endpoints/default/kubernetes". We'd fix this by writing
+multiple endpoints in there.
+
+#### 3. API server component Health check
+
+Every ApiServer will fetch the default kubernetes service/endpoints periodically
+and check each other's health, if health check failed, remove the failed endpoint
+from the EndpointSubset.
+
+#### 4. Client behavior
+
+1) Ramdomly select a API server endpoint
+
+By defaut, client of scheduler/controller/kubelet would accept a API server
+addresses list, and try to connect one, if connect succeed, fetch the default
+kubernetes service/endpoints. When a request failed, try to connect another
+endpoint. Request will failed iff all the API server endpoints failed.
```

Comment:
We should consider building this only in one or a few components. kube-proxy already does this random distribution; if we just tell components the service ip, since the endpoints are populated correctly already, nothing needs to be done.

We do need to teach either kubelet or kube-proxy to persist multiple apiserver endpoints, but that's so it can reconnect on reboot.